File: 백준/Gold/1987. 알파벳/알파벳.java

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.*;
import java.io.IOException;
import java.lang.StringBuilder;

class Main {

    static int[][] arr;
    static int[][] isvisited;
    static int[] dx = {0, 1, 0, -1};
    static int[] dy = {1, 0, -1, 0};
    static int n, m;
    static int[] alphabet = new int[26];
    static int maxCnt = 1;
    static int curCnt = 1;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] strs = br.readLine().split(" ");
        n = Integer.parseInt(strs[0]);
        m = Integer.parseInt(strs[1]);
        arr = new int[n][m];
        isvisited = new int[n][m];

        for(int i = 0; i < n; i++){
            String str = br.readLine();
            for(int j = 0; j < str.length(); j++){
                arr[i][j] = str.charAt(j) - 'A';
            }
        }

        // 0, 0을 지나야함.
        isvisited[0][0] = 1;
        alphabet[arr[0][0]] = 1;
        dfs(0, 0);

        System.out.println(maxCnt);
    }

    public static void dfs(int x, int y){

        for(int i = 0; i < 4; i++){
            int nx = x + dx[i];
            int ny = y + dy[i];

            if(nx < 0 || ny < 0 || nx >= n || ny >= m
                || isvisited[nx][ny] == 1 || alphabet[arr[nx][ny]] == 1) continue;

            alphabet[arr[nx][ny]] = 1;
            isvisited[nx][ny] = 1;
            curCnt++;
            maxCnt = Math.max(maxCnt, curCnt);

            dfs(nx, ny);
            alphabet[arr[nx][ny]] = 0;
            isvisited[nx][ny] = 0;
            curCnt--;
        }
    }

}

File: 백준/Gold/1987. 알파벳/알파벳.py
import sys
inp = sys.stdin.readline

# 세로 r칸 가로 c칸 -> r*c # 65, 90
n, m = map(int, inp().strip().split(" "))
arr = [[0] * m for _ in range(n)]
alphabet = [0] * 26
dx, dy = [0, 1, 0, -1],[1, 0, -1, 0]
maxx = -1
visit = [[0] * m for _ in range(n)]

for i in range(n):
    strr = inp().strip()
    for j in range(m):
        arr[i][j] = ord(strr[j]) - 65

def dfs(x, y, cnt):
    global maxx
    maxx = max(maxx, cnt)
    for i in range(4):
        nx = x + dx[i] 
        ny = y + dy[i]

        if nx < 0 or ny < 0 or nx >= n or ny >= m or visit[nx][ny] == 1: continue
        if alphabet[arr[nx][ny]] == 1: continue
        alphabet[arr[nx][ny]] = 1
        visit[nx][ny] = 1
        dfs(nx, ny, cnt+1)
        visit[nx][ny] = 0
        alphabet[arr[nx][ny]] = 0

alphabet[arr[0][0]] = 1
visit[0][0] = 1
dfs(0, 0, 1)

print(maxx)